openapi-converter/openapi_to_clients_schema/src/openapi.rs

```diff
@@ -0,0 +1,408 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! OpenAPI schema utilities
+
+use std::fmt::Debug;
+use openapiv3::*;
+use std::ops::Deref;
+use anyhow::{anyhow, bail};
+use serde_json::Value as JsonValue;
+use tracing::info;
+
+/// A wrapper around an openapi schema, also providing helper methods to explore it.
+pub struct OpenAPI(pub openapiv3::OpenAPI);
+
+impl Deref for OpenAPI {
+    type Target = openapiv3::OpenAPI;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl OpenAPI {
+
+    ///
+    /// Get the schema for a reference, if it exists, and follows references until we find a schema
+    ///
+    pub fn ref_to_schema(&self, reference: &str) -> anyhow::Result<&Schema> {
+
+        if let Some(name) = reference.strip_prefix("#/components/schemas/") {
+            match self.components.as_ref().unwrap().schemas.get(name) {
+                Some(ReferenceOr::Item(sch)) => return Ok(sch),
+                Some(ReferenceOr::Reference { reference }) => return self.ref_to_schema(reference),
+                _ => {},
+            }
+        }
+
+        bail!("No schema definition found for {reference}")
+    }
+
+    ///
+    /// Get the schema for a reference or schema
+    ///
+    pub fn get_schema<'a>(&'a self, r_or_s: &'a ReferenceOr<Schema>) -> anyhow::Result<&'a Schema> {
+        match r_or_s {
+            ReferenceOr::Reference { reference } => self.ref_to_schema(&reference),
+            ReferenceOr::Item(schema) => Ok(&schema),
+        }
+    }
+
+    ///
+    /// Is this type nullable?
+    ///
+    pub fn is_nullable(&self, id: &str) -> bool {
+        self.components
+            .as_ref()
+            .and_then(|c| c.schemas.get(id))
+            .map(|s| match s {
+                ReferenceOr::Reference { reference: _r } => false,
```

Comment:
Should this reference call recurse into the referenced type to determine if it's nullable?